># STGRNS
![image](https://github.com/zhanglab-wbgcas/STGRNS/blob/main/Figure%201.jpg)

># Introduction

STGRNS, a Transformer-based model, provides a fast and accurate  tool to infer gene regulatory networks from a single-cell RNA-seq profile. 
By leveraging the newly designed neural network structure, 
STGRNS especially obtains an outperformance on GRN inference. 

Instructions and examples are provided in the following tutorials.

># Requirement

- scikit-learn (Compatible with all versions)
- Pytorch (With Cudatoolkit is recommanded)
- Numpy > 1.20
- Scanpy > 1.9.1

># Data Availability
For the GRN reconstruction task, the processed experimental single-cell gene expression datasets are available on Zenodo at  https://doi.org/10.5281/zenodo.3378975.  
The gene expression and ChIP-Seq data of bone marrow-derived macrophages, dendritic cells, and mESC(1) are available at https://github.com/xiaoyeye/CNNC. 


[//]: # (```)

># Tutorial
- [Inferring gene regulatory networks](https://github.com/zhanglab-wbgcas/STGRNS/blob/main/Tutorial.ipynb)



># Paper Link
https://academic.oup.com/bioinformatics/article/39/4/btad165/7099621



